### PR TITLE
FLIP: Epochs dynamic node startup

### PR DESCRIPTION
**Pull request by kc1116**
## Description

FLIP describing mechanism to dynamically start node trying to join the network.

### CODE REVIEW

**Comment by kc1116** [comment]
This looks good, thanks for beefing it up @jordanschalm . We now need @vishalchangrani to take 1 more look .

File: flips/20220111-node-operator-network-joining-improvements.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+| **Updated**   | 2022-01-11                                           |
+
+## Objective
+
+Define the tools provided the node software to facilitate starting a newly staked node at the appropriate
+time during the epoch cycle.
+
+## Background
+
+Currently, node operators can only join or leave the network during network sporks. Epochs enable nodes to
```
Comment (vishalchangrani) [inline flips/20220111-node-operator-network-joining-improvements.md:17]:
```suggestion
Currently, node operators can **only** join or leave the network during network sporks. Epochs enable nodes to 
```

File: flips/20220111-node-operator-network-joining-improvements.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+We need a mechanism that will allow node operators to configure their nodes to bootstrap and join the network 
+at the appropriate epoch counter and phase. This mechanism will facilitate the process of joining the network 
+at epoch boundaries as opposed to when sporks happen.
+
+## User Benefit
+
+Node operators will be able to join the network at any epoch boundary after staking operations have completed
+without requiring manual intervention at the time when their node becomes eligible to participate in network communications.
+
```
Comment (vishalchangrani) [inline flips/20220111-node-operator-network-joining-improvements.md:67]: +Node operators do NOT have to wait for a spork to join the network.

**Review by vishalchangrani — APPROVED** [review]
❤️  thanks for putting this together

File: flips/20220111-node-operator-network-joining-improvements.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+| Status        | (Proposed / Rejected / Accepted / Implemented)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [752](https://github.com/onflow/flow/pull/752)       |
+| **Author(s)** | Khalil Claybon (khalil.claybon@dapperlabs.com), Jordan Schalm (jordan@dapperlabs.com)       |
+| **Sponsor**   | Khalil Claybon (khalil.claybon@dapperlabs.com), Jordan Schalm (jordan@dapperlabs.com)       |
+| **Updated**   | 2022-01-11                                           |
+
+## Objective
+
+Define the tools provided the node software to facilitate starting a newly staked node at the appropriate
```
Comment (jordanschalm) [inline flips/20220111-node-operator-network-joining-improvements.md:12]:
```suggestion
Define the tools provided by the node software to facilitate starting a newly staked node at the appropriate
```

File: flips/20220111-node-operator-network-joining-improvements.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+- Wait for specified epoch counter & phase.
+- Retrieve protocol root snapshot using flow client.
+- Continue with normal module initialization.
+
+The following CLI arguments will be added:
+- startup-epoch-counter: The epoch in which to start the node.
+- startup-epoch-phase: The phase within the specified epoch in which to start the node.
+- root-snapshot-dir: Directory where downloaded root snapshot should be saved when using
+  the dynamic join functionality.
+- dynamic-join: When enabled start node using mechanism described above.
```
Comment (jordanschalm) [inline flips/20220111-node-operator-network-joining-improvements.md:83]:
```suggestion
- `startup-epoch-counter`: The epoch in which to start the node.
- `startup-epoch-phase`: The phase within the specified epoch in which to start the node.
- `root-snapshot-dir`: Directory where downloaded root snapshot should be saved when using
  the dynamic join functionality.
- `dynamic-join`: When enabled start node using mechanism described above.
```

**Review by jordanschalm — APPROVED** [review]

**Comment by jordanschalm** [comment]
Relevant documentation is here: https://docs.onflow.org/node-operation/node-bootstrap/

**Pull request closed without merging**